docs/shim-rune.md

```diff
@@ -0,0 +1,241 @@
+# Design
+
+According to [Enclave Confidential Containers (Enclave-CC) Design and Architecture](./design.md),
+shim-rune should launch the agent enclave container when pod creation. The agent enclave container
+is deployed in the form of OCI bundle rather than a container image, and thus `shim-rune` can call
+`rune` to start it. The agent enclave container has the same life cycle as the pod. Its main job
+is to receive and then load the actual workload's container image into an app enclave.
+
+![image](./images/shim.png)
+
+As is shown in the picture, the binary name of `shim-rune` is `containerd-shim-rune-v2`.
+`shim-rune` should implement the Containerd Runtime V2 (Shim API) for Enclave-CC. With `shim-rune`,
+Kubernetes can launch Pod and OCI-compatible containers with one shim per Pod.
+
+## Goals
+
+- Implements [Containerd Runtime V2 (Shim API)](https://github.com/containerd/containerd/tree/main/runtime/v2)
+- Use `rune` OCI runtime to manage container
+- Launch agent enclave container during PodSandbox creation
+- Complete PullImage with agent enclave container with ttrpc communication
+- Launch application enclave container with FUSE encryption filesystem generated by agent enclave
+container.
+- Kill and cleanup agent enclave container resource when stopping PodSandbox
+
+# Workflow
+
+## 1. RunPodSandbox
+
+Compared with `containerd-shim-runc-v2`, `shim-rune` should add the following functions when
+launching the pod sandbox.
+
+- In the `Create` API of shim-rune, shim-rune will create an agent enclave container with the
+pre-created oci bundle after the pause container is created successfully.
+- In the `Start` API of `shim-rune, shim-rune will start the agent enclave container after the pause
+container is started successfully.
+
+A pre-created oci bundle for the agent enclave container is needed instead of the form of container
+image because the agent enclave container is the first component along with pod creation and there
+is no direct support for it to pull and unpack it in a TEE. The `agent_container_instance` field in
+the configuration file of shim-rune(In the Appendix section) shows the host path of oci bundle for
+the agent enclave container. The contents of the bundle are as follows:
+
+```sh
+tree /opt/enclave-cc/agent-instance/ -L 1
+/opt/enclave-cc/agent-instance/
+├── config.json
+└── rootfs
+```
+
+Note that oci bundle of the agent enclave container (`agent_container_instance` field in the
+shim-rune configuration file) must be **read-only** because the agent enclave containers of all
```

Comment:
Thank you for your suggestion. The config.json and rootfs in the bundle are read-only in our current design.

The pr-installed OCI bundle of agent enclave container will be shared on all pods in the host. Suppose the OCI bundle is in `/opt/enclave-cc/agent-instance/`. Using the same OCI bundle to run two agent enclave containers is impossible in Occlum Libos, because each occlum run will initialize the rootfs in the OCI bundle, and the second reinitialization will panic. 

After discussion, we provide the following solution:

Every time the agent container is created and started, `/opt/enclave-cc/agent-instance/` is used as the lower read-only layer, and then a read-write layer (upperdir) is superimposed through overlayfs, then the final OCI bundle for the agent container is aggregated.

As shown below, suppose the agent container  id is "a3e3c2"

```
mount -t overlay overlay -o
  lowerdir= /opt/enclave-cc/agent-instance/ ,
  upperdir=/run/containerd/agentenclave/a3e3c2/upper,
  workdir=/run/containerd/agentenclave/a3e3c2/work, 
  /run/containerd/agentenclave/a3e3c2/merged
```
Note that: `upperdir`, `mergeddir`, `workdir` are all created and managed by shim-rune.

Reply:
makes sense